### PR TITLE
replace reconcililation of ceph csi deployments with ceph-csi-operator resources

### PR DESCRIPTION
**Pull request by leelavg**
We are directly deploying and reconciling ceph csi deployments, driver, daemonset and configmap among other things and with ceph-csi-operator we can use it's APIs for delegating the reconciliation of these resources.

Loosely, the mapping is as follows:

csi deployments, daemonset, csidriver is handled by operatorconfig, driver APIs
csi configmap is split b/n cephcluster & config APIs providing a separation of reconciliation by different controllers

As a result of using APIs, the shape of data is changed particularly around supplying csi images (aka via imageset) and so an init container splits the current format into individual configmaps that can be provided as a references to csi-op APIs directly

Code path is separated as best as possible to eventually reconcile with only one way.

Pending:

- [x] olm dependency file for ceph-csi-operator.
- [x] kernelmountoptions facility in Config CRD
- [X] align the labels on CSI pods w/ controller & nodeplugin (maybe backward incompat)
- [X] align w/ SA defined in csi-op for SCC

Testing is done wherever possible.

pls note, the naming of new APIs is subject to change.

Depends on: red-hat-storage/ocs-operator#2713

part of [RHSTOR-5704]

### CODE REVIEW

File: pkg/templates/csi.go
Comment (Madhu-1) [inline pkg/templates/csi.go:92]: can you please revist this one again and compare it with existing deployment/daemonset to ensure we have everything what is required to be set here?

Comment (leelavg) [reply pkg/templates/csi.go:92]: I did refer the existing details before coming up w/ above. Now that Ohad mentioned that ability to add volumes is updated, I'll be refreshing that.

File: controllers/operatorconfigmap_controller.go
Comment (leelavg) [reply controllers/operatorconfigmap_controller.go:880]: done.

File: controllers/operatorconfigmap_controller.go
Comment (leelavg) [reply controllers/operatorconfigmap_controller.go:308]: done.

File: controllers/operatorconfigmap_controller.go
Comment (leelavg) [reply controllers/operatorconfigmap_controller.go]: done.

File: Dockerfile
Comment (leelavg) [reply Dockerfile]: done.

File: controllers/operatorconfigmap_controller.go
Comment (leelavg) [reply controllers/operatorconfigmap_controller.go:891]: that's the task of this function, get a matching imageset from a list.

**Comment by leelavg** [comment]
iirc, there might be some issue w/ toolchain in d/s, will re-look before merge.

**Comment by leelavg** [comment]
Update: No regressions found, old functionality is working as expected.

File: controllers/operatorconfigmap_controller.go
Comment (Madhu-1) [inline controllers/operatorconfigmap_controller.go:190]: This looks to be wider permission, we can settle on what is required?

File: controllers/storageclaim_controller.go
Comment (Madhu-1) [inline controllers/storageclaim_controller.go:267]: can you please move this change to a new commit?

File: controllers/storageclaim_controller.go
Comment (Madhu-1) [inline controllers/storageclaim_controller.go:525]: can we avoid this duplication and handle the config and CR part inside a if/else check which can minimise the code duplication?

File: controllers/storageclient_controller.go
Comment (Madhu-1) [inline controllers/storageclient_controller.go:255]: add a TODO to handle kernel mount options as well?

File: hack/go-build.sh
Comment (Madhu-1) [inline hack/go-build.sh:12]: lets update the build team for this new change

**Review by Madhu-1 — COMMENTED** [review]
LGTM, small nits, will review once we have other pending API implementation is done

File: controllers/storageclaim_controller.go
Comment (leelavg) [reply controllers/storageclaim_controller.go:267]: I observed it's taking more time to separate the commits rather than actually working on the commit or raising a PR. Raised a new PR along with some backlog items.

File: controllers/operatorconfigmap_controller.go
Comment (leelavg) [reply controllers/operatorconfigmap_controller.go:190]: 2 reasons I didn't do that yet, one except the mapping API we need all permissions for every other API, two not directly related, I'm waiting for names to be finalized.

File: controllers/storageclaim_controller.go
Comment (leelavg) [reply controllers/storageclaim_controller.go:525]: not possible at the moment, every where I'm using big branching rather than minor variants w/ li'l copying because at the end only one survives before going to QE.

There were some implicit assumptions in existing code on how the type of storageclass is found out, I don't see a reason to follow that.

If it's not mandatory, I'd want to defer this comment.

Comment (leelavg) [reply controllers/storageclaim_controller.go]: done.

File: service/csi-image-set-creator/main.go
Comment (nb-ohad) [inline service/csi-image-set-creator/main.go:1]: Let's not create runtime code that compensates for the lack of a build time code that we control!

File: service/status-report/main.go
Comment (nb-ohad) [inline service/status-report/main.go:181]: This need to be handled by the operator not the reporter

**Review by nb-ohad — CHANGES_REQUESTED** [review]

File: bundle/manifests/ocs-client-operator.clusterserviceversion.yaml
Comment (leelavg) [reply bundle/manifests/ocs-client-operator.clusterserviceversion.yaml:183]: ack, got a similar comment earlier, wanted to do it after the names are changed, anyways will do it.

File: bundle/manifests/ocs-client-operator.clusterserviceversion.yaml
Comment (leelavg) [reply bundle/manifests/ocs-client-operator.clusterserviceversion.yaml:774]: we talked about it couple of times, until the new op stabilizes I don't want to go to build team and revert the changes. As of now, build team packs all the images into a single config (as usual) and we are splitting that in init container.

File: controllers/storageclaim_controller.go
Comment (leelavg) [reply controllers/storageclaim_controller.go:414]: > the resource type is not a "Spec"
- ack, until names are changed I'm just being explicit.

> I would also think we should start to consider sending
- basically the runtime.Object? Should we directly send it or add new specific fields?

File: controllers/storageclaim_controller.go
Comment (leelavg) [reply controllers/storageclaim_controller.go:415]: I am divided here, I can remove this if we are guaranteed to have csi-op bundled.

File: controllers/storageclaim_controller.go
Comment (leelavg) [reply controllers/storageclaim_controller.go:417]: yes, only the resource name comes from provider (here, it's `cephfs` or `ceph-rbd`) but if there are more than single claim the names could clash and so attaching the existing hash here.

we could also concatenate storagerequest hash to the resource and directly send from provider as well.

File: controllers/operatorconfigmap_controller.go
Comment (leelavg) [reply controllers/operatorconfigmap_controller.go:335]: yes, as per https://docs.google.com/document/d/10kib7JQAKf2oXLHjquEAxNyvQASitW9m2aDs3j6iB5Y/edit?disco=AAABO7dTLaE

File: controllers/operatorconfigmap_controller.go
Comment (leelavg) [reply controllers/operatorconfigmap_controller.go:345]: sure

File: controllers/operatorconfigmap_controller.go
Comment (leelavg) [reply controllers/operatorconfigmap_controller.go:355]: yes

File: controllers/operatorconfigmap_controller.go
Comment (leelavg) [reply controllers/operatorconfigmap_controller.go:368]: yes

File: controllers/operatorconfigmap_controller.go
Comment (leelavg) [reply controllers/operatorconfigmap_controller.go:377]: no, in 4.16 the deployment & daemonset names are different, presence of csi-op doesn't automatically delete the existing resources and we don't want two resources reconciling same drivers.

File: controllers/operatorconfigmap_controller.go
Comment (leelavg) [reply controllers/operatorconfigmap_controller.go:587]: I am divided here. I tried that once and I moved a lot of code more than what I did here and so left it out. I don't think doing it twice calls for a good amount of refactor and specifically for the code that should be deleted.

File: controllers/operatorconfigmap_controller.go
Comment (leelavg) [reply controllers/operatorconfigmap_controller.go:924]: > Let's break it into 2 steps
- ack, is straightforward.

> you'd have no reason to do
I need to know the metadata of configmap to decide whether that corresponds to the platform or not

File: controllers/operatorconfigmap_controller.go
Comment (leelavg) [reply controllers/operatorconfigmap_controller.go:963]: > You should assume a new configmap
- not yet, same as https://github.com/red-hat-storage/ocs-client-operator/pull/175#discussion_r1692578565

File: pkg/templates/csi.go
Comment (leelavg) [reply pkg/templates/csi.go:99]: ack, in first commit it had context around it and I explicitly made a commit to directly use driver spec from csi-op in second commit and so these values were left behind.

File: pkg/utils/k8sutils.go
Comment (leelavg) [reply pkg/utils/k8sutils.go:137]: this just refactors a function which was used in two places and I didn't add anything here, can remove it from the PR.

File: service/csi-image-set-creator/main.go
Comment (leelavg) [reply service/csi-image-set-creator/main.go:1]: will revisit after https://github.com/red-hat-storage/ocs-client-operator/pull/175#discussion_r1692578565

File: service/status-report/main.go
Comment (leelavg) [reply service/status-report/main.go:181]: will revisit after https://github.com/red-hat-storage/ocs-client-operator/pull/175#discussion_r1692578565

File: controllers/storageclaim_controller.go
Comment (nb-ohad) [reply controllers/storageclaim_controller.go:424]: No, I am not talking about changing the data we send. The data is either the spec of the resource or the data section of a config map or a secret. I am talking about the identifier of the kind of the object. Today we send only Kind, but this is problematic when you have more than one resource using the same Kind string. I am suggesting to ass the group and version to kind so we could fully identify by VGK

File: controllers/storageclaim_controller.go
Comment (nb-ohad) [reply controllers/storageclaim_controller.go:425]: I think you misunderstood me. I am saying that the delegate decision should come from the provider side. By choosing to send config (ClientProfile) spec it indicates that it wants you to delegate CSI

File: controllers/storageclaim_controller.go
Comment (nb-ohad) [reply controllers/storageclaim_controller.go:427]: What am I missing, each storage claim has its own config (client profile) which is named with the claim name. Claims are cluster scoped so we cannot have 2 claims with the same name.

File: controllers/operatorconfigmap_controller.go
Comment (nb-ohad) [reply controllers/operatorconfigmap_controller.go:335]: But if it is the same clustername why do we even need to set it?  how would that help CSI differentiate the drivers?

File: controllers/operatorconfigmap_controller.go
Comment (nb-ohad) [reply controllers/operatorconfigmap_controller.go:377]: Delete operation should only be sent to the API server once. If the config is not a good indicator let's find another one. But we should not send delete operations on each reconcile.

File: controllers/storageclaim_controller.go
Comment (leelavg) [reply controllers/storageclaim_controller.go:425]: Then we want to make provider also aware of how client want to deploy CSI.

Pls do note that, even before we get first response from provider we'll deploy csi pods and I don't think provider need to explicitly know to use csi-op or not as the decision is already taken.

File: controllers/storageclaim_controller.go
Comment (leelavg) [reply controllers/storageclaim_controller.go:427]: Legit, a fundamental mistake. Will change.

File: controllers/operatorconfigmap_controller.go
Comment (leelavg) [reply controllers/operatorconfigmap_controller.go:335]: @Madhu-1 possible to mention where this clusterID be used?

File: controllers/operatorconfigmap_controller.go
Comment (leelavg) [reply controllers/operatorconfigmap_controller.go:377]: Two options:
1. Store the result in reconciler
2. Get before Delete which hits cache before hitting API

File: controllers/operatorconfigmap_controller.go
Comment (leelavg) [reply controllers/operatorconfigmap_controller.go:588]: The initial comment is on deletionPhase, so do you want the decision happen once and also fold the deletion into the same?

File: controllers/storageclaim_controller.go
Comment (nb-ohad) [reply controllers/storageclaim_controller.go:425]: @leelavg you are generalizing. This code path can never run before we have a response from the provider and at that point the `DeployCSI` check is meaningless and we should remove the check. 

I was not talking or referring to any other place in the code base where we are using `DeployCSI`

Comment (leelavg) [reply controllers/storageclaim_controller.go:425]: when `DelegateCSI` is set, then it is an indication we'll have csi-op CRDs in the cluster, if we don't check that, creation of the `Config CR` (older terms) will fail.

File: controllers/operatorconfigmap_controller.go
Comment (nb-ohad) [reply controllers/operatorconfigmap_controller.go:377]: @leelavg I don't mind having a couple of deletes just because of cache delay. But not having delete ops sent to the k8s API on every reconcile

Comment (leelavg) [reply controllers/operatorconfigmap_controller.go:377]: pardon, is the last sentence cut off? I couldn't understand the intention.

Comment (nb-ohad) [reply controllers/operatorconfigmap_controller.go:377]: Just make sure we are not sending continues unnecessary delete operations to the k8s API Server.

File: pkg/templates/csi.go
Comment (leelavg) [reply pkg/templates/csi.go]: done.

File: controllers/operatorconfigmap_controller.go
Comment (leelavg) [reply controllers/operatorconfigmap_controller.go]: done.

File: controllers/storageclaim_controller.go
Comment (leelavg) [reply controllers/storageclaim_controller.go]: already merged. can be resolved.

File: bundle/manifests/ocs-client-operator.clusterserviceversion.yaml
Comment (leelavg) [reply bundle/manifests/ocs-client-operator.clusterserviceversion.yaml]: done.

File: controllers/storageclaim_controller.go
Comment (leelavg) [reply controllers/storageclaim_controller.go]: done.

File: controllers/storageclaim_controller.go
Comment (leelavg) [reply controllers/storageclaim_controller.go]: done.

File: controllers/operatorconfigmap_controller.go
Comment (leelavg) [reply controllers/operatorconfigmap_controller.go]: done.

File: controllers/operatorconfigmap_controller.go
Comment (leelavg) [reply controllers/operatorconfigmap_controller.go]: > Let's break it into 2 steps
- done something similar

File: pkg/templates/csi.go
Comment (leelavg) [reply pkg/templates/csi.go]: done.

File: controllers/storageclaim_controller.go
Comment (leelavg) [reply controllers/storageclaim_controller.go]: As discussed offline, provider sends both the older & newer resources w/o distinguishing client is delegating CSI and so client needs to decide whether to create the resource.

File: service/status-report/main.go
Comment (leelavg) [reply service/status-report/main.go]: https://github.com/red-hat-storage/ocs-client-operator/pull/175#discussion_r1706417411

File: hack/go-build.sh
Comment (nb-ohad) [reply hack/go-build.sh:13]: Sorry, I am still confused. Why are we pushing code that needs to happen in build time into runtime? 
I asked on multiple reviews to have the config maps be produced at build time, there is no point in having a container do a build time operation at runtime.

File: internal/controller/storageclaim_controller.go
Comment (nb-ohad) [inline internal/controller/storageclaim_controller.go:438]: This PR already established a pattern where the unmarshaling happens outside and before `createOrUpdate`, can you please align here as well?

File: internal/controller/storageclient_controller.go
Comment (nb-ohad) [inline internal/controller/storageclient_controller.go:232]: This PR already established a pattern where the unmarshaling happens outside and before createOrUpdate, can you please align here as well?

File: internal/controller/storageclient_controller.go
Comment (nb-ohad) [inline internal/controller/storageclient_controller.go:283]: Can you please specialize this comment based on the result instead of having a generic message:
```suggestion
	r.Log.Info(fmt.Sprintf("%s successfully %s", obj.GetKind, result), "name", obj.GetName())
```

File: pkg/utils/k8sutils.go
Comment (nb-ohad) [inline pkg/utils/k8sutils.go:1]: The new additions here are not generic k8s concerns and should be moved to a different util file.

File: pkg/utils/k8sutils.go
Comment (nb-ohad) [inline pkg/utils/k8sutils.go:119]: Can we have this as a function to prevent code from changing the value from outside this package?

File: service/status-report/main.go
Comment (nb-ohad) [inline service/status-report/main.go:181]: This code should not be part of the reporter. Please move it into the appropriate controller

**Review by nb-ohad — CHANGES_REQUESTED** [review]

File: hack/go-build.sh
Comment (leelavg) [reply hack/go-build.sh:13]: as commented in https://github.com/red-hat-storage/ocs-client-operator/pull/175#discussion_r1692578565

If we are not going to use csi-op then we need to revert the d/s build time code. Nevertheless, I'll ask build time to have both types of ConfigMaps be bundled.

File: internal/controller/storageclaim_controller.go
Comment (leelavg) [reply internal/controller/storageclaim_controller.go:438]: We are unmarshaling two shapes, one is of map[string]string and another is the spec specific to the resource that we are interested and I couldn't club both.

File: internal/controller/storageclient_controller.go
Comment (leelavg) [reply internal/controller/storageclient_controller.go:232]: We are unmarshaling two shapes, one is of map[string]string and another is the spec specific to the resource that we are interested and I couldn't club both.

File: service/status-report/main.go
Comment (leelavg) [reply service/status-report/main.go:181]: We need to update endpoints received as a response to heartbeat when the hash feature is in then controller will take over.

File: pkg/utils/k8sutils.go
Comment (leelavg) [reply pkg/utils/k8sutils.go]: done.

File: internal/controller/storageclient_controller.go
Comment (leelavg) [reply internal/controller/storageclient_controller.go]: done.

File: service/csi-image-set-creator/main.go
Comment (leelavg) [reply service/csi-image-set-creator/main.go:1]: done.

File: bundle/manifests/ocs-client-operator.clusterserviceversion.yaml
Comment (leelavg) [reply bundle/manifests/ocs-client-operator.clusterserviceversion.yaml]: done.

**Comment by nb-ohad** [comment]
@leelavg I think there is a vendor issue here. It should not be 414 files for just bringing the new operator's APIs package

**Pull request merged**
